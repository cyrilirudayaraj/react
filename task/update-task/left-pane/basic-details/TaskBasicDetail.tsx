import {
  Checkbox,
  Form,
  Accordion,
  AccordionItem,
  StatusTag,
} from '@athena/forge';
import React from 'react';
import Labels from '../../../../../constants/Labels';
import DecisionTable from '../decision-table/DecisionTable';
import TaskDependency from '../task-dependency/TaskDependency';
import { TaskBasicDetailProps } from '../../UpdateTaskProps';
import TaskTestingDetail from '../testing-details/TaskTestingDetail';
import WFTextarea from '../../../../../components/wf-textarea/WFTextarea';
import { isTaskEditable } from '../../UpdateTask';
import Acl from '../../../../../constants/Acl';
import WFEnableForPermission from '../../../../../components/wf-enableforpermission/WFEnableForPermission';
import RichTextEditor from '../../../../../components/rich-text-editor/RichTextEditor';
import AppConstants from '../../../../../constants/AppConstants';

export default function TaskBasicDetail(
  props: TaskBasicDetailProps
): JSX.Element {
  const labels = Labels.TASK_BASIC_DETAILS;
  const { formik } = props;
  const isDisabled = !isTaskEditable(props.task);
  const values = props.task?.taskDependencies;
  const actualDependencies = values?.filter(function (taskDependency) {
    return !taskDependency.deleted;
  });
  const totalDependencies = actualDependencies?.length;
  let completedDependenciesCount = 0;
  let pendingDependenciesCount = 0;
  actualDependencies?.forEach(function (taskDependency: any) {
    if (
      taskDependency.completedYn === 'Y' ||
      taskDependency.statusName ===
        AppConstants.UI_CONSTANTS.TASK_DEPENDENCIES.DEPENDENCY_BLOCKING_STATUS
          .COMPLETED
    ) {
      completedDependenciesCount = completedDependenciesCount + 1;
    } else {
      pendingDependenciesCount = pendingDependenciesCount + 1;
    }
  });
  function getOverallDependencyStatus(): string {
    return (
      completedDependenciesCount +
      '/' +
      totalDependencies +
      AppConstants.UI_CONSTANTS.WHITE_SPACE +
      AppConstants.UI_CONSTANTS.COMPLETE
    );
  }
  const getStatusCss = (): string => {
    let style = '';
    if (pendingDependenciesCount === 0 && completedDependenciesCount > 0) {
      style += ' completed';
    }
    return style;
  };
  return (
    <Form
      labelAlwaysAbove={true}
      includeSubmitButton={false}
      className="task-detail"
      requiredVariation="blueBarWithLegend"
    >
      <div className="panel">
        <div className="heading">
          <h1 className="content">
            {Labels.TASK_DETAIL_FORM.TASK_INFORMATION}
          </h1>
        </div>
        <div className="panel-content">
          <WFEnableForPermission permission={Acl.TASK_NAME_UPDATE}>
            <WFTextarea
              {...props}
              label={labels.TASK_NAME}
              name="name"
              maxlength={400}
              istextarea={false}
              disabled={isDisabled}
            />
          </WFEnableForPermission>
          <WFEnableForPermission permission={Acl.TASK_DESCRIPTION_UPDATE}>
            <RichTextEditor
              id="description"
              wrapperId={1}
              labelText={labels.TASK_DESCRIPTION}
              formik={formik}
              isDisabled={isDisabled}
              isRequired={props?.mandatoryFields['description'] || false}
            ></RichTextEditor>
          </WFEnableForPermission>
          <WFEnableForPermission
            permission={Acl.TASK_PRODUCTION_CLAIM_EXAMPLE_UPDATE}
          >
            <WFTextarea
              {...props}
              label={labels.CLAIM_EXAMPLE}
              name="testClaimExample"
              maxlength={200}
              istextarea={false}
              disabled={isDisabled}
            />
          </WFEnableForPermission>
        </div>
      </div>
      <Accordion className="taskgroup dependency">
        <AccordionItem
          headingText={
            Labels.TASK_DEPENDENCY_DETAILS.RELATED_TASKS_AND_DEPENDENCIES
          }
          defaultExpanded={true}
          headerSlot={
            <StatusTag
              className={`status_nowrap_no_transform ${getStatusCss()}`}
              text={getOverallDependencyStatus()}
            />
          }
        >
          <TaskDependency {...props}></TaskDependency>
        </AccordionItem>
      </Accordion>
      <Accordion className="taskgroup">
        <AccordionItem
          headingText={Labels.TASK_DETAIL_FORM.CHANGES}
          defaultExpanded={true}
        >
          <WFEnableForPermission
            permission={Acl.TASK_MODEL_DESIGN_CHANGES_UPDATE}
          >
            <RichTextEditor
              id="modelDesignChanges"
              wrapperId={2}
              labelText={labels.MODEL_DESIGN_CHANGES}
              formik={formik}
              isDisabled={isDisabled}
              isRequired={props?.mandatoryFields['modelDesignChanges'] || false}
            ></RichTextEditor>
          </WFEnableForPermission>
          <DecisionTable {...props}></DecisionTable>
          <div className="fe_c_form-field requires_checkbox">
            <span>Requires:</span>
            <WFEnableForPermission
              permission={Acl.TASK_REQUIRES_ANET_CHANGES_UPDATE}
            >
              <Checkbox
                description={labels.ATHENANET_CHANGES}
                id="athenaNetChangesYn"
                checked={formik.values.athenaNetChangesYn}
                {...formik.getFieldProps('athenaNetChangesYn')}
                disabled={isDisabled}
              >
                {/* <React.Fragment /> */}
              </Checkbox>
            </WFEnableForPermission>
            <WFEnableForPermission
              permission={Acl.TASK_REQUIRES_DT_CHANGES_UPDATE}
            >
              <Checkbox
                description={labels.ODM_CHANGES}
                id="dtChangesYn"
                checked={formik.values.dtChangesYn}
                {...formik.getFieldProps('dtChangesYn')}
                disabled={isDisabled}
              >
                {/* <React.Fragment /> */}
              </Checkbox>
            </WFEnableForPermission>
          </div>
        </AccordionItem>
      </Accordion>
      <Accordion className="taskgroup">
        <AccordionItem
          headingText={Labels.TASK_DETAIL_FORM.TESTING}
          defaultExpanded={true}
        >
          <TaskTestingDetail {...props} />
        </AccordionItem>
      </Accordion>
    </Form>
  );
}
