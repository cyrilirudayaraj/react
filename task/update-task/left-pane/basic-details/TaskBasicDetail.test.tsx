import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TaskBasicDetail from './TaskBasicDetail';

configure({ adapter: new Adapter() });
let wrapper: any;
const formik: any = {
  values: {
    athenaNetChangesYn: false,
    businessRequirementDesc: '<p>test</p>',
    businessRequirementId: '338',
    businessRequirementName:
      'Force Drop Specified Policy/Group Numbers for ECTE',
    businessRequirementTypeId: '1',
    context: '',
    description: '<p>test</p>',
    dtBranchUrl: '',
    dtChangesYn: false,
    id: '150',
    contextId: null,
    localRuleUseCaseId: null,
    modelBranchUrl: '',
    modelDesignChanges: '<p>test</p>',
    name: 'test',
    ruleReportingCategoryId: '13',
    taskDecisionTableDetails: [],
    taskDependencies: [],
    testClaimExample: 'test',
    testNote: '',
    version: '10',
  },
  touched: {},
  errors: {},
  getFieldProps: jest.fn(() => {
    return { value: 'test' };
  }),
};
const mandatoryFields: any = {
  description: true,
  modelDesignChanges: true,
  name: true,
  testClaimExample: true,
};

const task: any = {
  activeTaskStepId: '1047',
  activeTaskStepName: 'Review Task',
  activeWorkflowStepId: '2',
  assignedTo: 'bvenkatesan',
  associatedBRTasks: [],
  athenaNetChangesYn: 'N',
  businessRequirementDesc: '<p>test</p>',
  businessRequirementId: '338',
  businessRequirementName: 'Force Drop Specified Policy/Group Numbers for ECTE',
  businessRequirementTypeId: '1',
  canSignOff: '1',
  clientDueDate: null,
  created: '02/28/2021',
  createdBy: 'jcyril',
  deploymentDate: null,
  description: '<p>test</p>',
  dtBranchUrl: null,
  dtChangesYn: 'N',
  dueDate: '03/02/2021',
  escalatedYn: null,
  id: '150',
  jiraId: null,
  lastModified: '02/28/2021',
  lastModifiedBy: 'jcyril',
  legacyRuleArchived: null,
  legacyRuleId: '1.2129',
  legacyTaskId: '12',
  contextId: null,
  contextName: null,
  contextRefId: null,
  localRuleUseCaseId: null,
  modelBranchUrl: null,
  modelDesignChanges: '<p>test</p>',
  name: 'test',
  originatingCaseId: '12',
  originatingSystemId: '1',
  originatingSystemName: 'Salesforce',
  priorityId: '1',
  priorityName: '0',
  ruleReportingCategoryId: '13',
  ruleReportingCategoryName: 'Validation: Patient Insurance Issues',
  status: 'Assigned',
  statusId: '3',
  taskDecisionTableDetails: [],
  taskDependencies: [],
  taskSteps: [
    {
      actualCompletion: '02/28/2021',
      actualStart: '02/28/2021',
      assignedTo: 'jcyril',
      id: '1046',
      initialCompletion: '02/28/2021',
      initialStart: '02/28/2021',
      name: 'Analyze Task',
      ordering: '1',
      plannedCompletion: null,
      plannedStart: null,
      signOffStatement: 'My analysis is complete',
      signedOffYn: 'Y',
      totalWorkLog: '45',
      transitionText: 'Send for Analysis',
      version: '9',
      workflowStepId: '1',
    },
  ],
  taskTypeId: '3',
  taskTypeName: 'Dual Update',
  testClaimExample: 'test',
  testNote: null,
  version: '10',
};

beforeEach(() => {
  wrapper = shallow(
    <TaskBasicDetail
      formik={formik}
      task={task}
      mandatoryFields={mandatoryFields}
    />
  );
});
describe('TaskBasicDetail', () => {
  it('basic snapshot', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
