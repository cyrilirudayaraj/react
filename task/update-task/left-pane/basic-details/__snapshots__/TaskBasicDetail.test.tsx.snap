// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskBasicDetail basic snapshot 1`] = `
"<Form labelAlwaysAbove={true} includeSubmitButton={false} className=\\"task-detail\\" requiredVariation=\\"blueBarWithLegend\\" buttonText=\\"Submit\\" noValidate={true} layout=\\"medium\\" nested={true} labelWidth={3}>
  <div className=\\"panel\\">
    <div className=\\"heading\\">
      <h1 className=\\"content\\">
        Task Information
      </h1>
    </div>
    <div className=\\"panel-content\\">
      <Connect(WFEnableForPermission) permission=\\"task.name.update\\">
        <WFTextarea formik={{...}} task={{...}} mandatoryFields={{...}} label=\\"Task Name\\" name=\\"name\\" maxlength={400} istextarea={false} disabled={false} />
      </Connect(WFEnableForPermission)>
      <Connect(WFEnableForPermission) permission=\\"task.description.update\\">
        <Connect(RichTextEditor) id=\\"description\\" wrapperId={1} labelText=\\"Task Description\\" formik={{...}} isDisabled={false} isRequired={true} />
      </Connect(WFEnableForPermission)>
      <Connect(WFEnableForPermission) permission=\\"task.production-claim-example.update\\">
        <WFTextarea formik={{...}} task={{...}} mandatoryFields={{...}} label=\\"Production Claim Example\\" name=\\"testClaimExample\\" maxlength={200} istextarea={false} disabled={false} />
      </Connect(WFEnableForPermission)>
    </div>
  </div>
  <Accordion className=\\"taskgroup dependency\\">
    <AccordionItem headingText=\\"Related Tasks and Dependencies\\" defaultExpanded={true} headerSlot={{...}} headingTag=\\"h3\\" mountedWhileHidden={false} padded={true} tabIndex={0}>
      <Connect(TaskDependency) formik={{...}} task={{...}} mandatoryFields={{...}} />
    </AccordionItem>
  </Accordion>
  <Accordion className=\\"taskgroup\\">
    <AccordionItem headingText=\\"Changes\\" defaultExpanded={true} headingTag=\\"h3\\" mountedWhileHidden={false} padded={true} tabIndex={0}>
      <Connect(WFEnableForPermission) permission=\\"task.model-design-changes.update\\">
        <Connect(RichTextEditor) id=\\"modelDesignChanges\\" wrapperId={2} labelText=\\"Model Design Changes\\" formik={{...}} isDisabled={false} isRequired={true} />
      </Connect(WFEnableForPermission)>
      <DecisionTable formik={{...}} task={{...}} mandatoryFields={{...}} />
      <div className=\\"fe_c_form-field requires_checkbox\\">
        <span>
          Requires:
        </span>
        <Connect(WFEnableForPermission) permission=\\"task.requires-anet-changes.update\\">
          <Checkbox description=\\"athenaNet Changes\\" id=\\"athenaNetChangesYn\\" checked={false} value=\\"test\\" disabled={false} hideRequiredStyles={false} required={false} />
        </Connect(WFEnableForPermission)>
        <Connect(WFEnableForPermission) permission=\\"task.requires-dt-changes.update\\">
          <Checkbox description=\\"ODM Changes\\" id=\\"dtChangesYn\\" checked={false} value=\\"test\\" disabled={false} hideRequiredStyles={false} required={false} />
        </Connect(WFEnableForPermission)>
      </div>
    </AccordionItem>
  </Accordion>
  <Accordion className=\\"taskgroup\\">
    <AccordionItem headingText=\\"Testing\\" defaultExpanded={true} headingTag=\\"h3\\" mountedWhileHidden={false} padded={true} tabIndex={0}>
      <TaskTestingDetail formik={{...}} task={{...}} mandatoryFields={{...}} />
    </AccordionItem>
  </Accordion>
</Form>"
`;
